### PR TITLE
TouchID Implementation [iOS Sim Only]

### PR DESCRIPTION
**Pull request by dpgraham**
## Change list

Added touchId command
 
## Types of changes

What types of changes are you proposing/introducing to Java client?
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Details

Touch ID is a feature in Apple products (see https://support.apple.com/en-ca/HT201371). This endpoint simulates the behaviour.

### CODE REVIEW

**Comment by dpgraham** [comment]
*Not sure why the README.md was diffed that way. I only added one line under the 'Added Functions' header

File: src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java [hunk 1]
```diff
@@ -31,8 +34,6 @@
 
 import java.io.File;
 
-import static org.junit.Assert.assertEquals;
-
 public class XCUIAutomationTest {
```
Comment (SrinivasanTarget) [inline src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java]: @dpgraham Can you add some tests here?

File: src/main/java/io/appium/java_client/ios/IOSMobileCommandHelper.java [hunk 2]
```diff
@@ -81,4 +81,15 @@
         return new AbstractMap.SimpleEntry<>(
                 SHAKE, ImmutableMap.<String, Object>of());
     }
+    
+    /**
+     * This method forms a {@link java.util.Map} of parameters for the touchId simulator
+     * 
+     * @param match Are we simulating a successful fingerprint scan?
```
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/ios/IOSMobileCommandHelper.java]: Why is it a question here?

File: src/main/java/io/appium/java_client/ios/IOSDeviceActionShortcuts.java [hunk 3]
```diff
@@ -56,5 +57,14 @@ default void hideKeyboard(String strategy, String keyName) {
     default void shake() {
         CommandExecutionHelper.execute(this, shakeCommand());
     }
+    
+    /**
+     * Simulate touchId event
+     *
+     * @param match Are we simulating a successful fingerprint scan?
+     */
+    default void touchId(boolean match) {
```
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/ios/IOSDeviceActionShortcuts.java]: Can you add in doc it is Simulator only?

File: README.md [hunk 4]
```diff
@@ -1,311 +1,312 @@
...
+- zoom()
+- isAppInstalled()
+- installApp()
+- removeApp()
+- launchApp()
+- closeApp()
+- endTestCoverage()
+- isLocked()
+- shake()
+- touchId()
```
Comment (SrinivasanTarget) [inline README.md]: Can you revert this change? We will apply this change post release?

File: src/main/java/io/appium/java_client/ios/IOSMobileCommandHelper.java (hunk 2 above)
Comment (dpgraham) [inline src/main/java/io/appium/java_client/ios/IOSMobileCommandHelper.java]: It's how I've documented boolean parameters in the past. I'll change it to be more in line with how it's normally documented.

File: src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java (hunk 1 above)
Comment (dpgraham) [inline src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java]: I added the test. It doesn't do any assertions because there isn't anything to check, it just calls the methods to verify that there aren't any exceptions.

**Comment by dpgraham** [comment]
Thanks @SrinivasanTarget, I made the changes you requested. If you could review them and they're good let me know so that I can squash the changes and merge them.

File: src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java [hunk 5]
```diff
@@ -91,4 +92,14 @@
...
         assertEquals(driver.rotation(), landscapeLeftRotation);
     }
+
+    @Test public void testTouchId() {
+        try {
+            driver.touchId(true);
+            driver.touchId(false);
+            assertEquals(true, true);
+        } catch(Exception e){
+            assertEquals(true, false);
```
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java]: @dpgraham I think it is more senseful to throw an exception.
Otherwice we have a risk to produce and invalid build or to not react to server changes.

File: src/main/java/io/appium/java_client/ios/IOSDeviceActionShortcuts.java [hunk 6]
```diff
@@ -56,5 +57,14 @@ default void hideKeyboard(String strategy, String keyName) {
     default void shake() {
         CommandExecutionHelper.execute(this, shakeCommand());
     }
+
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/ios/IOSDeviceActionShortcuts.java]: I am against the adding to the `IOSDeviceActionShortcuts`. This interface is going to be deprecated. Please take a look [at this working branch](https://github.com/TikhomirovSergey/java-client/tree/SrinivasanTarget-TouchActionSplitting/src/main/java/io/appium/java_client/ios). [IOSDeviceActionShortcuts](https://github.com/TikhomirovSergey/java-client/blob/SrinivasanTarget-TouchActionSplitting/src/main/java/io/appium/java_client/ios/IOSDeviceActionShortcuts.java) is going to be deprecated. 

It is better to add such interface:

```java
package io.appium.java_client.ios;

public interface ChecksTouchId extends ExecutesMethod {

    /**
    * Simulate touchId event
    *
    * @param match If true, simulates a successful fingerprint scan. If false, simulates a failed fingerprint scan
    */     
    default void touchId(boolean match) {
        CommandExecutionHelper.execute(this, touchIdCommand(match));
    }
}
```

and make IOSDriver implement it.

**Comment by TikhomirovSergey** [comment]
@dpgraham
Sorry for the delaying. Usually busy. Please take a look at remarks.

File: src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java [hunk 7]
```diff
@@ -91,4 +92,14 @@
         driver.rotate(landscapeLeftRotation);
         assertEquals(driver.rotation(), landscapeLeftRotation);
     }
+
+    @Test public void testTouchId() {
+        try {
+            driver.touchId(true);
+            driver.touchId(false);
+            assertEquals(true, true);
```
Comment (SrinivasanTarget) [inline src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java]: Actual Usecase of API ```touchID``` is to unlock or sign in to any application using touchID. Since we dont have our test app built with any registration or sign in feature, we cannot use that scenario.

Otherwise Can we lock the simulator and unlock the simulator using ```driver.touchID(true)```.Ideal case this should be working fine.

File: src/main/java/io/appium/java_client/ios/IOSDeviceActionShortcuts.java [hunk 8]
```diff
@@ -56,5 +57,14 @@ default void hideKeyboard(String strategy, String keyName) {
     default void shake() {
         CommandExecutionHelper.execute(this, shakeCommand());
     }
+    
+    /**
+     * Simulate touchId event
+     *
+     * @param match If true, simulates a successful fingerprint scan. If false, simulates a failed fingerprint scan
```
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/ios/IOSDeviceActionShortcuts.java]: If i'm not wrong, this will fail as per google style each doc line should end with a period.

File: src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java (hunk 7 above)
Comment (SrinivasanTarget) [inline src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java]: So test scenario should be something like,

```
   try {
            driver.lockDevice(2);
            driver.touchId(true);
        } catch (Exception e) {
            assertTrue(false);
        }
```

**Comment by SrinivasanTarget** [comment]
@dpgraham I've reviewed again now. Please take a look.

File: src/main/java/io/appium/java_client/ios/IOSDeviceActionShortcuts.java (hunk 6 above)
Comment (dpgraham) [inline src/main/java/io/appium/java_client/ios/IOSDeviceActionShortcuts.java]: Thanks for the input Sergey. I made the change.

**Comment by dpgraham** [comment]
I've updated the code again

File: src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java [hunk 9]
```diff
@@ -91,4 +92,15 @@
...
         assertEquals(driver.rotation(), landscapeLeftRotation);
     }
+
+    @Test public void testTouchId() {
+        try {
+            driver.performTouchID(true);
+            driver.performTouchID(false);
+            assertEquals(true, true);
+        } catch (Exception e) {
+            assertEquals(true, false);
```
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java]: Test should throw the exception. 

At least you can do that 

```java
try {
//do something
//assert the result
} catch (Exception e) {
    throw e;
} finally {
//attemt to stabilize the simulator
//or something else
}
```

File: src/main/java/io/appium/java_client/ios/IOSMobileCommandHelper.java [hunk 10]
```diff
@@ -81,4 +81,15 @@
         return new AbstractMap.SimpleEntry<>(
                 SHAKE, ImmutableMap.<String, Object>of());
     }
+    
+    /**
+     * This method forms a {@link java.util.Map} of parameters for the touchId simulator
+     * 
+     * @param match If true, simulates a successful fingerprint scan. If false, simulates a failed fingerprint scan
```
Comment (SrinivasanTarget) [inline src/main/java/io/appium/java_client/ios/IOSMobileCommandHelper.java]: Doc should end with a period everywhere.

**Comment by dpgraham** [comment]
Changes made

**Comment by TikhomirovSergey** [comment]
@dpgraham Everuthing is ok for me. But could please improve the proposed test the way I have asked?
@SrinivasanTarget ?

File: src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java (hunk 9 above)
Comment (SrinivasanTarget) [inline src/test/java/io/appium/java_client/ios/XCUIAutomationTest.java]: @dpgraham Can you update tests here as suggested please?
